ansible_builder/containerfile.py

```diff
@@ -0,0 +1,365 @@
+import importlib.resources
+import logging
+import os
+
+from pathlib import Path
+
+from . import constants
+from .user_definition import UserDefinition
+from .utils import copy_directory, copy_file
+
+
+logger = logging.getLogger(__name__)
+
+
+class Containerfile:
+    newline_char = '\n'
+
+    def __init__(self, definition: UserDefinition,
+                 build_context=None,
+                 container_runtime=None,
+                 output_filename=None,
+                 galaxy_keyring=None,
+                 galaxy_required_valid_signature_count=None,
+                 galaxy_ignore_signature_status_codes=()):
+        """
+        :param str galaxy_keyring: GPG keyring file used by ansible-galaxy to opportunistically validate collection signatures.
+        :param str galaxy_required_valid_signature_count: Number of sigs (prepend + to disallow no sig) required for ansible-galaxy to accept collections.
+        :param str galaxy_ignore_signature_status_codes: GPG Status codes to ignore when validating galaxy collections.
+        """
+
+        self.build_context = build_context
+        self.build_outputs_dir = os.path.join(
+            build_context, constants.user_content_subfolder)
+        self.definition = definition
+        if output_filename is None:
+            filename = constants.runtime_files[container_runtime]
+        else:
+            filename = output_filename
+        self.path = os.path.join(self.build_context, filename)
+        self.container_runtime = container_runtime
+        self.original_galaxy_keyring = galaxy_keyring
+        self.copied_galaxy_keyring = None
+        self.galaxy_required_valid_signature_count = galaxy_required_valid_signature_count
+        self.galaxy_ignore_signature_status_codes = galaxy_ignore_signature_status_codes
+        self.steps: list = []
+
+    def prepare(self):
+        """
+        Prepares the steps for the run-time specific build file.
+
+        Incrementally builds the `self.steps` attribute by extending it with the
+        info to eventually be written directly to the container definition file
+        via a separate call to the `Containerfile.write()` method.
+        """
+
+        # Build args all need to go at top of file to avoid errors
+        self._insert_global_args(include_values=True)
+
+        ######################################################################
+        # Zero stage: prep base image
+        ######################################################################
+
+        # 'base' (possibly customized) will be used by future build stages
+        self.steps.extend([
+            "# Base build stage",
+            "FROM $EE_BASE_IMAGE as base",
+        ])
```

Comment:
Should we include a `USER root` here?  There is no guarantee that the `base_image` will use a default user of `root`.  An example image `registry.access.redhat.com/ubi8/python-39:latest` uses a default non-root user, of `1001`, and the below `dnf` commands fail with:

```
Error: This command has to be run with superuser privileges (under the root user on most systems).
Error: building at STEP "RUN dnf install $PYPKG -y && dnf clean all": while running runtime: exit status 1
```

Reply:
https://github.com/nitzmahone/ansible-builder/pull/12